### PR TITLE
feat: add `queryInterface.withoutForeignKeyChecks`, `sequelize.destroyAll`

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [ ] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

I needed a way to delete all rows from all tables between tests for https://github.com/sequelize/sequelize/pull/15840. `TRUNCATE` is not a viable solution when the table is referenced by other tables (even if they're empty) in some dialects.

This PR:

- Adds the ability to skip foreign key checks in `sequelize.truncate` for mysql, sqlite, mariadb via the a option: `withoutForeignKeyChecks`
- Adds a new method on `queryInterface`: `withoutForeignKeyChecks`, which can be used to temporarily disable foreign keys for a given connection.
- Adds `sequelize.destroyAll` which can be used as an alternative to `sequelize.truncate`, as it can handle foreign keys.
- Replaces internal uses of a few deprecated APIs, like `Model.tableName`
- Makes `transaction.connection` hard private with a getter
- Updates `setTransactionFromCls` to make it synchronise the `transaction` and `connection` parameters when possible. This makes designing APIs that use these two easier. You can now specify both as long as they are compatible.
- Replaces our test utils `disableDatabaseResetForSuite` & `enableTruncateDatabaseForSuite` with `setResetMode`

## TODO

- [x] add tests
- [x] change the suite reset hook to have one hook that accepts multiple modes: `delete-schema`, `truncate`, `destroy` and `none`
- [x] Implement https://github.com/sequelize/sequelize/issues/15388

### CODE REVIEW

File: packages/core/test/integration/support.ts [hunk 1]
```diff
@@ -1,19 +1,17 @@
-import assert from 'node:assert';
 import pTimeout from 'p-timeout';
 import { QueryTypes } from '@sequelize/core';
 import type { AbstractQuery } from '@sequelize/core/_non-semver-use-at-your-own-risk_/dialects/abstract/query.js';
 import * as Support from '../support';
```
Comment (WikiRik) [inline packages/core/test/integration/support.ts:4]: Not really in the scope of this PR, but can we only import the methods we're actually using here?

File: packages/core/test/integration/support.ts [hunk 2]
```diff
@@ -1,19 +1,17 @@
...
 
 // Mocha still relies on 'this' https://github.com/mochajs/mocha/issues/2657
```
Comment (WikiRik) [inline packages/core/test/integration/support.ts:6]: Do we still need this comment here now that the rule is not disabled?

File: packages/core/test/integration/model/internals.test.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
+import { expect } from 'chai';
+import type { Transactionable } from '@sequelize/core';
+import { setTransactionFromCls } from '@sequelize/core/_non-semver-use-at-your-own-risk_/model-internals.js';
+import { beforeAll2, createSequelizeInstance, getTestDialect } from '../../support';
+
+const dialectName = getTestDialect();
+
+describe('setTransactionFromCls', () => {
+  const vars = beforeAll2(async () => {
+    const sequelize = await createSequelizeInstance({
```
Comment (WikiRik) [inline packages/core/test/integration/model/internals.test.ts:10]:
```suggestion
    const sequelize = createSequelizeInstance({
```

File: packages/core/src/transaction.ts [hunk 4]
```diff
@@ -193,12 +205,13 @@ export class Transaction {
...
     // if we've already cleaned up
-    if (this.parent || this.connection?.uuid === undefined) {
+    if (this.parent || this.#connection?.uuid === undefined) {
       return;
     }
 
-    this.sequelize.connectionManager.releaseConnection(this.connection);
-    this.connection.uuid = undefined;
+    this.sequelize.connectionManager.releaseConnection(this.#connection);
+    this.#connection.uuid = undefined;
```
Comment (WikiRik) [inline packages/core/src/transaction.ts:213]: I don't think we need to set this to undefined if `this.#connection` itself is undefined in the following line

File: packages/core/src/transaction.ts [hunk 5]
```diff
@@ -210,12 +223,13 @@ export class Transaction {
...
-    if (this.parent || this.connection?.uuid === undefined) {
+    if (this.parent || this.#connection?.uuid === undefined) {
       return;
     }
 
-    await this.sequelize.connectionManager.destroyConnection(this.connection);
-    this.connection.uuid = undefined;
+    await this.sequelize.connectionManager.destroyConnection(this.#connection);
+    this.#connection.uuid = undefined;
+    this.#connection = undefined;
```
Comment (WikiRik) [inline packages/core/src/transaction.ts:232]: Same here

File: packages/core/src/dialects/sqlite/query-generator-typescript.ts [hunk 6]
```diff
@@ -19,6 +19,10 @@ export class SqliteQueryGeneratorTypeScript extends MySqlQueryGenerator {
     return `PRAGMA INDEX_LIST(${this.quoteTable(tableName)})`;
   }
 
+  getToggleForeignKeyChecksQuery(enable: boolean): string {
```
Comment (WikiRik) [inline packages/core/src/dialects/sqlite/query-generator-typescript.ts:22]: Can we add unit tests for the generation of this query?

File: packages/core/src/transaction.ts (hunk 4 above)
Comment (ephys) [reply packages/core/src/transaction.ts:213]: Yes but the connection still exists, it's returned to the pool

File: packages/core/src/dialects/sqlite/query-generator-typescript.ts (hunk 6 above)
Comment (ephys) [reply packages/core/src/dialects/sqlite/query-generator-typescript.ts:22]: Done

File: packages/core/test/integration/model/internals.test.ts (hunk 3 above)
Comment (ephys) [reply packages/core/test/integration/model/internals.test.ts]: Done

File: packages/core/test/integration/support.ts (hunk 2 above)
Comment (ephys) [reply packages/core/test/integration/support.ts]: Removed

File: packages/core/test/integration/support.ts (hunk 1 above)
Comment (ephys) [reply packages/core/test/integration/support.ts]: Done

**Review by WikiRik — APPROVED** [review]

**Pull request merged**